CodingStandard.md

```diff
@@ -0,0 +1,1425 @@
+CURIOUS(!is_crc_valid(d), "Invalid CRC");  // logs but does not terminate
+return false;  // the correctness of the software must not depend
+               // on whether is_crc_valid(d) is executed
+```
+
+**Logging**
+
+Logging macros should also be created and applied generously throughout the code
+to capture system state and behavior information.
+
+Guidelines for Object-Oriented and Related Technologies
+================================================================================
+
+**Implementation Inheritance**
+
+<u>Summary</u>
+
+- Prefer composition over implementation inheritance for code reuse
+- Prefer private or protected inheritance for implementation inheritance
+- Do not redefine non-virtual members of a base class
+
+<u>Discussion</u>
+
+Implementation inheritance occurs whenever a base class having state data (non-
+static member variables) or implemented member functions (either virtual or
+non-virtual) is inherited by a derived class.  In particular, it is not
+considered implementation inheritance if the base class contains no state data
+and only pure-virtual member function declarations (no implementations).
+
+In most cases, implementation inheritance should be expressed with private or
+protected inheritance.  Public implementation inheritance can be dangerous
+because the requirements of the base class may be overlooked by users of the
+derived class.  As a result, the base class members could be put into an
+invalid state.
+
+Keeping implementation inheritance protected or private constrains the use of
+the base class, precluding substitution, so users of the derived class need
+be concerned only with the requirements of its public interface.  It is then
+the responsibility of the derived class implementation to ensure the base
+class requirements are met.
+
+Composition (putting common code into a class that is reused as a member of
+other classes) is generally preferable to implementation inheritance for ease
+of maintainability.  Before choosing implementation inheritance, decide if
+composition is a suitable alternative.  [^Sutter34]
+
+If implementation inheritance is used, ensure that none of the non-virtual
+base class member functions are redefined by the derived class.  Also ensure
+that the correct base class constructor is called whenever a derived class
+constructor is invoked.
+
+Implementation inheritance relationships should be explained in the in the
+`@inherits` field of the derived class header.
+
+**Interface Inheritance**
+
+<u>Summary</u>
+
+- Public inheritance implies interface inheritance and represents an "is-a"
+  relationship
+- All publicly derived classes must meet the requirements of the base class
+- Interface inheritance must satisfy the Liskov Substitution Principle
+- Avoid mixing interface inheritance and implementation inheritance if possible
+- Consider using abstract base classes, the non-virtual interface idiom, or the
+  curiously recurring template parameter idiom
```

Comment:
Let's talk about these idioms, I haven't heard them outside this document

Reply:
I admit I'm not really familiar with them either.  I'll figure out what they're talking about or possibly just remove this section.